File: src/types/plan.ts
export interface Options {
	stopSteam: boolean
	restartSteam: boolean
	replaceExistingArtwork: boolean
	createCollections: boolean
}

export interface LauncherSettings {
	enabled: boolean
	customPath: string | null
}

export interface SteamGridDbSettings {
	enabled: boolean
	apiKey: string | null
}

export interface PersistedSettings {
	stopSteam: boolean
	restartSteam: boolean
	createCollections: boolean
	steamLocation: string | null
	launchers: Record<string, LauncherSettings>
	steamGridDb: SteamGridDbSettings
}

export interface PreviewPlan {
	userSteamId: string
	changes: PlannedChange[]
	filesToChange: string[]
	backups: BackupPlan[]
	requiresSteamRestart: boolean
}

export interface BackupPlan {
	source: string
	destination: string
}

import type { ArtworkKind, ArtworkSource } from './import'

export interface PlannedChange {
	id: string
	gameName: string
	file: string
	kind: 'addShortcut' | 'updateShortcut' | 'writeArtwork' | 'updateCollections'
	destructive: boolean
	artworkSource: ArtworkSource | null
	artworkKind: ArtworkKind | null
	collectionName: string | null
}

export interface ApplyResult {
	appliedChanges: PlannedChange[]
	backupsCreated: string[]
}

export interface BackupInfo {
	id: string
	createdAt: string
	fileCount: number
	sizeBytes: number
}

File: src-tauri/src/models/plan.rs
use super::importers::{ArtworkKind, ArtworkSource, ImportCandidate};
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PreviewPlan {
    pub user_steam_id: String,
    pub changes: Vec<PlannedChange>,
    pub files_to_change: Vec<PathBuf>,
    pub backups: Vec<BackupPlan>,
    pub requires_steam_restart: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BackupPlan {
    pub source: PathBuf,
    pub destination: PathBuf,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlannedChange {
    pub id: String,
    pub game_name: String,
    pub file: PathBuf,
    pub kind: ChangeKind,
    pub destructive: bool,
    pub artwork_source: Option<ArtworkSource>,
    pub artwork_kind: Option<ArtworkKind>,
    pub collection_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ChangeKind {
    AddShortcut,
    UpdateShortcut,
    WriteArtwork,
    UpdateCollections,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyRequest {
    pub plan: PreviewPlan,
    pub candidates: Vec<ImportCandidate>,
    pub options: Options,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Options {
    pub stop_steam: bool,
    pub restart_steam: bool,
    pub replace_existing_artwork: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyResult {
    pub applied_changes: Vec<PlannedChange>,
    pub backups_created: Vec<PathBuf>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BackupInfo {
    pub id: String,
    pub created_at: String,
    pub file_count: usize,
    pub size_bytes: u64,
}
